discoverd/server/store.go

```diff
@@ -0,0 +1,907 @@
+	}
+	sub.el = s.subscribers[service].PushBack(sub)
+
+	// Send current instances.
+	if sendCurrent && kinds.Any(discoverd.EventKindUp) {
+		for _, inst := range s.instances(service) {
+			ch <- &discoverd.Event{
+				Service:  service,
+				Kind:     discoverd.EventKindUp,
+				Instance: inst,
+			}
+			// TODO: add a timeout to sends so that clients can't slow things down too much
+		}
+	}
+
+	// Send current leader.
+	if leader := s.leader(service); sendCurrent && kinds&discoverd.EventKindLeader != 0 && leader != nil {
+		ch <- &discoverd.Event{
+			Service:  service,
+			Kind:     discoverd.EventKindLeader,
+			Instance: leader,
+		}
+	}
+
+	// Send current service meta data.
+	if meta := s.serviceMeta(service); sendCurrent && kinds.Any(discoverd.EventKindServiceMeta) && meta != nil {
+		ch <- &discoverd.Event{
+			Service:     service,
+			Kind:        discoverd.EventKindServiceMeta,
+			ServiceMeta: meta,
+		}
+	}
+
+	// Send current service.
+	if sendCurrent && kinds.Any(discoverd.EventKindCurrent) {
+		ch <- &discoverd.Event{
+			Service: service,
+			Kind:    discoverd.EventKindCurrent,
+		}
+	}
+
+	return sub
+}
+
+// Broadcast sends an event to all subscribers.
+func (s *Store) Broadcast(event *discoverd.Event) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.broadcast(event)
+}
+
+func (s *Store) broadcast(event *discoverd.Event) {
+	// Retrieve list of subscribers for the service.
+	l, ok := s.subscribers[event.Service]
+	if !ok {
+		return
+	}
+
+	// Iterate over each subscriber in the list.
+	for el := l.Front(); el != nil; el = el.Next() {
+		sub := el.Value.(*subscription)
+
+		// Skip if event type is not subscribed to.
+		if sub.kinds&event.Kind == 0 {
+			continue
+		}
+
+		// Send event to subscriber.
+		// If subscriber is blocked then close it.
+		select {
+		case sub.ch <- event:
+		default:
+			sub.err = ErrSendBlocked
+			go sub.Close()
+		}
+	}
+}
+
+// raftSnapshot implements raft.FSMSnapshot.
+// The FSM is serialized on snapshot creation so this simply writes to the sink.
+type raftSnapshot struct {
+	data []byte
+}
+
+// Persist writes the snapshot to the sink.
+func (ss *raftSnapshot) Persist(sink raft.SnapshotSink) error {
+	// Write data to sink.
+	if _, err := sink.Write(ss.data); err != nil {
+		sink.Cancel()
+		return err
+	}
+
+	// Close and exit.
+	return sink.Close()
+}
+
+// Release implements raft.FSMSnapshot. This is a no-op.
+func (ss *raftSnapshot) Release() {}
+
+// instanceEntry represents an instance with a TTL.
+type instanceEntry struct {
+	Instance   *discoverd.Instance
+	ExpiryTime time.Time
+}
+
+// instanceSlice represents a sortable list of instances by id.
+type instanceSlice []*discoverd.Instance
+
+func (a instanceSlice) Len() int           { return len(a) }
+func (a instanceSlice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a instanceSlice) Less(i, j int) bool { return a[i].ID < a[j].ID }
+
+// Command type header bytes.
+const (
+	addServiceCommandType     = byte(0)
```

Comment:
I prefer to hardcode the values when they'll be persisted to disk. Otherwise it's too easy to insert a value in the middle and corrupt the other values.